### PR TITLE
feat: Support retrying complex workflows with nested group nodes

### PR DESCRIPTION
**Pull request by terrytangyuan**
This includes:
* Major refactoring with clearer logic.
* Connected parent group nodes are reset correctly.
* Only the correct branch that matches the selected node path in the DAG is retried.
* Handle pod deletion properly and fix some cases where nodes fail with "pod deleted" error.
* Added extensive test cases to cover retrying on any selected node in the workflow.

Signed-off-by: Yuan Tang <terrytangyuan@gmail.com>

### CODE REVIEW

File: workflow/util/util.go [hunk 1]
```diff
@@ -813,6 +847,7 @@ func FormulateRetryWorkflow(ctx context.Context, wf *wfv1.Workflow, restartSucce
 	deletedNodes := make(map[string]bool)
 	deletedPods := make(map[string]bool)
 	var podsToDelete []string
+	var resettedParentGroupNodes []string
```
Comment (juliev0) [inline workflow/util/util.go:850]: sorry to be annoying but can this be changed to "resetParentGroupNodes" since the past tense verb is "reset"?

Comment (juliev0) [reply workflow/util/util.go:850]: (same for all uses of this word)

Comment (terrytangyuan) [reply workflow/util/util.go:850]: Definitely! I’ll update this. Thanks!

File: workflow/util/util.go [hunk 2]
```diff
@@ -765,6 +765,40 @@ func deletePodNodeDuringRetryWorkflow(wf *wfv1.Workflow, node wfv1.NodeStatus, d
...
+		if !containsNode(resettedParentGroupNodes, currentNodeID) {
+			newWF.Status.Nodes[currentNodeID] = resetNode(*currentNode.DeepCopy())
+			resettedParentGroupNodes = append(resettedParentGroupNodes, currentNodeID)
+			log.Debugf("Resetting connected group node %s", currentNode.Name)
+		}
+		if currentNode.BoundaryID != "" && currentNode.BoundaryID != oldWF.ObjectMeta.Name {
+			parentNode := oldWF.Status.Nodes[currentNode.BoundaryID]
+			if isGroupNode(parentNode) {
+				currentNodeID = parentNode.ID
+			}
```
Comment (juliev0) [inline workflow/util/util.go:794]: What happens in the implicit else case here? If I'm reading it correctly, it seems currentNodeID doesn't get changed and we end up in an infinite loop, no? Or am I mistaken?

File: workflow/util/util.go [hunk 3]
```diff
@@ -821,37 +856,62 @@ func FormulateRetryWorkflow(ctx context.Context, wf *wfv1.Workflow, restartSucce
...
 						}
+						if nodeGroupNeedsReset {
+							newWF, resettedParentGroupNodes = resetConnectedParentGroupNodes(wf, newWF, node, resettedParentGroupNodes)
+						}
 					}
 				} else {
+					if node.Type == wfv1.NodeTypePod || node.Type == wfv1.NodeTypeSuspend {
+						newWF, resettedParentGroupNodes = resetConnectedParentGroupNodes(wf, newWF, node, resettedParentGroupNodes)
+						// Only remove the descendants of a suspended node but not the suspended node itself. The descendants
+						// of a suspended node need to be removed since because the conditions should be re-evaluated based on
```
Comment (juliev0) [inline workflow/util/util.go:883]: typo in comment: change "since because" to either "since" or "because"

**Review by juliev0 — CHANGES_REQUESTED** [review]
Requested changes (inline)

**Comment by terrytangyuan** [comment]
@juliev0 Thanks for your feedback! All comments have been addressed.

**Review by juliev0 — APPROVED** [review]

**Comment by terrytangyuan** [comment]
Test failure is unrelated: https://github.com/argoproj/argo-workflows/issues/9591

**Review by rbreeze — APPROVED** [review]
LGTM!

**Pull request merged**